AssertionError: Latent size(last 3 dimensions) should be divisible by patch size([1, 2, 2]), but got torch.Size([1, 45, 80]).

An invalid image generation dimension has an odd exception raised. Maybe it is correct, but the error message could be more helpful.

It took me a second to track down that that I had added the wrong height in my prompt. Initially I found #13 and was searching for a corrupt cached latent. The error message does not make it apparent that it was due to the prompt.

I had:

```
--w 640 --h 360 --d 1024 --g 5 --s 20 --f 1 --fs 3
```

Update to the following corrected the error:

```
--w 640 --h 480 --d 1024 --g 5 --s 20 --f 1 --fs 3
```

<details>
<summary>error</summary>

```shell
INFO     __main__ | prompt: this prompt errors (sample_image_inference)                            hv_train_network.py:987
INFO     __main__ | height: 360 (sample_image_inference)                                           hv_train_network.py:988
INFO     __main__ | width: 640 (sample_image_inference)                                            hv_train_network.py:989
INFO     __main__ | frame count: 1 (sample_image_inference)                                        hv_train_network.py:990
INFO     __main__ | sample steps: 20 (sample_image_inference)                                      hv_train_network.py:991
INFO     __main__ | guidance scale: 5.0 (sample_image_inference)                                   hv_train_network.py:992
INFO     __main__ | discrete flow shift: 3.0 (sample_image_inference)                              hv_train_network.py:993
INFO     __main__ | seed: 1024 (sample_image_inference)                                            hv_train_network.py:995
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ ...\musubi-tuner-scripts\musubi-tuner\hv_train_network.py:2708 in <module>  │
│                                                                                                  │
│   2705 │   args = parser.parse_args()                                                            │
│   2706 │   args = read_config_from_file(args, parser)                                            │
│   2707 │   trainer = NetworkTrainer()                                                            │
│ ❱ 2708 │   trainer.train(args)                                                                   │
│   2709                                                                                           │
│                                                                                                  │
│ ...\musubi-tuner-scripts\musubi-tuner\hv_train_network.py:1883 in train     │
│                                                                                                  │
│   1880 │   │   # For --sample_at_first                                                           │
│   1881 │   │   if should_sample_images(args, global_step, epoch=0):                              │
│   1882 │   │   │   optimizer_eval_fn()                                                           │
│ ❱ 1883 │   │   │   self.sample_images(accelerator, args, 0, global_step, vae, transformer, samp  │
│   1884 │   │   │   optimizer_train_fn()                                                          │
│   1885 │   │   if len(accelerator.trackers) > 0:                                                 │
│   1886 │   │   │   # log empty object to commit the sample images to wandb                       │
│                                                                                                  │
│ ...\musubi-tuner-scripts\musubi-tuner\hv_train_network.py:917 in            │
│ sample_images                                                                                    │
│                                                                                                  │
│    914 │   │   │   # If only one device is available, just use the original prompt list. We don  │
│    915 │   │   │   with torch.no_grad(), accelerator.autocast():                                 │
│    916 │   │   │   │   for sample_parameter in sample_parameters:                                │
│ ❱  917 │   │   │   │   │   self.sample_image_inference(                                          │
│    918 │   │   │   │   │   │   accelerator, args, transformer, dit_dtype, vae, save_dir, sample  │
│    919 │   │   │   │   │   )                                                                     │
│    920 │   │   │   │   │   clean_memory_on_device(accelerator.device)                            │
│                                                                                                  │
│ ...\musubi-tuner-scripts\musubi-tuner\hv_train_network.py:1009 in           │
│ sample_image_inference                                                                           │
│                                                                                                  │
│   1006 │   │   │   logger.info(f"control video path: {control_video_path}")                      │
│   1007 │   │                                                                                     │
│   1008 │   │   # inference: architecture dependent                                               │
│ ❱ 1009 │   │   video = self.do_inference(                                                        │
│   1010 │   │   │   accelerator,                                                                  │
│   1011 │   │   │   args,                                                                         │
│   1012 │   │   │   sample_parameter,                                                             │
│                                                                                                  │
│ ...\musubi-tuner-scripts\musubi-tuner\hv_train_network.py:1246 in           │
│ do_inference                                                                                     │
│                                                                                                  │
│   1243 │   │   guidance_expand = torch.tensor([guidance_scale * 1000.0], dtype=torch.float32, d  │
│   1244 │   │                                                                                     │
│   1245 │   │   # Get rotary positional embeddings                                                │
│ ❱ 1246 │   │   freqs_cos, freqs_sin = get_rotary_pos_embed_by_shape(transformer, latents.shape[  │
│   1247 │   │   freqs_cos = freqs_cos.to(device=device, dtype=dit_dtype)                          │
│   1248 │   │   freqs_sin = freqs_sin.to(device=device, dtype=dit_dtype)                          │
│   1249                                                                                           │
│                                                                                                  │
│ ...\musubi-tuner-scripts\musubi-tuner\hunyuan_model\models.py:1051 in       │
│ get_rotary_pos_embed_by_shape                                                                    │
│                                                                                                  │
│   1048 │   │   )                                                                                 │
│   1049 │   │   rope_sizes = [s // model.patch_size for s in latents_size]                        │
│   1050 │   elif isinstance(model.patch_size, list):                                              │
│ ❱ 1051 │   │   assert all(s % model.patch_size[idx] == 0 for idx, s in enumerate(latents_size))  │
│   1052 │   │   │   f"Latent size(last {ndim} dimensions) should be divisible by patch size({mod  │
│   1053 │   │   │   f"but got {latents_size}."                                                    │
│   1054 │   │   )                                                                                 │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
AssertionError: Latent size(last 3 dimensions) should be divisible by patch size([1, 2, 2]), but got torch.Size([1, 45,
80]).
```
</details>

--- a/hv_train_network.py
+++ b/hv_train_network.py
@@ -954,6 +954,10 @@ class NetworkTrainer:
         cfg_scale = sample_parameter.get("cfg_scale", None)  # None for architecture default
         negative_prompt = sample_parameter.get("negative_prompt", None)
 
+        # round width and height to multiples of 8
+        width = (width // 8) * 8
+        height = (height // 8) * 8
+
         frame_count = (frame_count - 1) // 4 * 4 + 1  # 1, 5, 9, 13, ... For HunyuanVideo and Wan2.1
 
         if self.i2v_training:

